File: tests/primitives/int_operators/int_operators.c
#include <stdio.h>
#include <stdlib.h>
#include <milone.h>

void int_operators_int_operators_literalTest(void);

void int_operators_int_operators_hexLiteralTest(void);

void int_operators_int_operators_arithmeticOperatorsTest(int32_t two_, int32_t three_, int32_t thirtyNine_);

void int_operators_int_operators_bitOperatorsTest(int32_t n1_);

void int_operators_int_operators_compareTest(int32_t n2_1, int32_t n3_);

void int_operators_int_operators_toIntTest(void);

int main(int argc, char **argv);

void int_operators_int_operators_literalTest(void) {
    if (((-(((int32_t)0x80000000 + 1))) != 2147483647)) milone_assert_error("int_operators/int_operators.milone", 4, 2);
    return;
}

void int_operators_int_operators_hexLiteralTest(void) {
    if (((int32_t)0x00000000 != 0)) milone_assert_error("int_operators/int_operators.milone", 7, 2);
    if (((int32_t)0x10 != 16)) milone_assert_error("int_operators/int_operators.milone", 8, 2);
    if (((int32_t)0x7fffffff != 2147483647)) milone_assert_error("int_operators/int_operators.milone", 9, 2);
    if (((int32_t)0xffffffff != -1)) milone_assert_error("int_operators/int_operators.milone", 10, 2);
    return;
}

void int_operators_int_operators_arithmeticOperatorsTest(int32_t two_, int32_t three_, int32_t thirtyNine_) {
    if (((two_ + three_) != 5)) milone_assert_error("int_operators/int_operators.milone", 13, 2);
    if (((three_ - two_) != 1)) milone_assert_error("int_operators/int_operators.milone", 14, 2);
    if (((two_ * three_) != 6)) milone_assert_error("int_operators/int_operators.milone", 15, 2);
    if (((thirtyNine_ / two_) != 19)) milone_assert_error("int_operators/int_operators.milone", 16, 2);
    if (((thirtyNine_ / three_) != 13)) milone_assert_error("int_operators/int_operators.milone", 17, 2);
    if (((thirtyNine_ % two_) != 1)) milone_assert_error("int_operators/int_operators.milone", 18, 2);
    if (((thirtyNine_ % three_) != 0)) milone_assert_error("int_operators/int_operators.milone", 19, 2);
    if (((-(two_)) != -2)) milone_assert_error("int_operators/int_operators.milone", 21, 2);
    return;
}

void int_operators_int_operators_bitOperatorsTest(int32_t n1_) {
    int32_t n2_;
    int32_t n4_;
    int32_t n8_;
    int32_t n16_;
    n2_ = (n1_ * 2);
    n4_ = (n2_ * 2);
    n8_ = (n4_ * 2);
    n16_ = (n8_ * 2);
    if (((((n1_ + n2_) + n4_) & ((n2_ + n4_) + n8_)) != (n2_ + n4_))) milone_assert_error("int_operators/int_operators.milone", 29, 2);
    if (((((n1_ + n2_) + n4_) | ((n2_ + n4_) + n8_)) != (((n1_ + n2_) + n4_) + n8_))) milone_assert_error("int_operators/int_operators.milone", 30, 2);
    if (((((n1_ + n2_) + n4_) ^ ((n2_ + n4_) + n8_)) != (n1_ + n8_))) milone_assert_error("int_operators/int_operators.milone", 31, 2);
    if (((n1_ << 0) != n1_)) milone_assert_error("int_operators/int_operators.milone", 33, 2);
    if (((n1_ << 1) != n2_)) milone_assert_error("int_operators/int_operators.milone", 34, 2);
    if (((((n1_ + n2_) + n8_) << 1) != ((n2_ + n4_) + n16_))) milone_assert_error("int_operators/int_operators.milone", 35, 2);
    if (((n4_ >> 0) != n4_)) milone_assert_error("int_operators/int_operators.milone", 37, 2);
    if (((n4_ >> 1) != n2_)) milone_assert_error("int_operators/int_operators.milone", 38, 2);
    if (((n4_ >> 2) != n1_)) milone_assert_error("int_operators/int_operators.milone", 39, 2);
    if (((n4_ >> 3) != 0)) milone_assert_error("int_operators/int_operators.milone", 40, 2);
    if (((((n1_ + n2_) + n8_) >> 1) != ((0 + n1_) + n4_))) milone_assert_error("int_operators/int_operators.milone", 41, 2);
    return;
}

void int_operators_int_operators_compareTest(int32_t n2_1, int32_t n3_) {
    if ((n2_1 != 2)) milone_assert_error("int_operators/int_operators.milone", 44, 2);
    if ((n3_ == 2)) milone_assert_error("int_operators/int_operators.milone", 45, 2);
    if ((2 < n2_1)) milone_assert_error("int_operators/int_operators.milone", 47, 2);
    if ((n3_ < n2_1)) milone_assert_error("int_operators/int_operators.milone", 48, 2);
    if ((n2_1 >= n3_)) milone_assert_error("int_operators/int_operators.milone", 49, 2);
    if ((n2_1 >= n3_)) milone_assert_error("int_operators/int_operators.milone", 51, 2);
    if ((milone_int32_compare(n2_1, 3) >= 0)) milone_assert_error("int_operators/int_operators.milone", 53, 2);
    if ((milone_int32_compare(n2_1, 2) != 0)) milone_assert_error("int_operators/int_operators.milone", 54, 2);
    if ((0 >= milone_int32_compare(n3_, 2))) milone_assert_error("int_operators/int_operators.milone", 55, 2);
    return;
}

void int_operators_int_operators_toIntTest(void) {
    int32_t call_;
    int32_t call_1;
    if ((0 != 0)) milone_assert_error("int_operators/int_operators.milone", 58, 2);
    if ((((int32_t)'a') != 97)) milone_assert_error("int_operators/int_operators.milone", 59, 2);
    call_ = string_to_int32((struct String){.ptr = "42", .len = 2});
    if ((call_ != 42)) milone_assert_error("int_operators/int_operators.milone", 60, 2);
    call_1 = string_to_int32((struct String){.ptr = "-1", .len = 2});
    if (((call_1 + 1) != 0)) milone_assert_error("int_operators/int_operators.milone", 61, 2);
    return;
}

int main(int argc, char **argv) {
    milone_start(argc, argv);
    int_operators_int_operators_literalTest();
    int_operators_int_operators_hexLiteralTest();
    int_operators_int_operators_arithmeticOperatorsTest(2, 3, 39);
    int_operators_int_operators_bitOperatorsTest(1);
    int_operators_int_operators_compareTest(2, 3);
    int_operators_int_operators_toIntTest();
    return 0;
}

File: tests/primitives/int_operators/int_operators.fs
module rec int_operators.Program

let literalTest () =
  // Int minimum/maximum values can be written as literal.
  assert (-((-2147483648) + 1) = 2147483647)

let hexLiteralTest () =
  assert (0x00000000 = 0)
  assert (0x10 = 16)
  assert (0x7fffffff = 2147483647)
  assert (0xffffffff = -1)

let arithmeticOperatorsTest (two: int) (three: int) (thirtyNine: int) =
  assert (two + three = 5)
  assert (three - two = 1)
  assert (two * three = 6)
  assert (thirtyNine / two = 19)
  assert (thirtyNine / three = 13)
  assert (thirtyNine % two = 1)
  assert (thirtyNine % three = 0)

  assert (-two = -2)

let bitOperatorsTest (n1: int) =
  let n2 = n1 * 2
  let n4 = n2 * 2
  let n8 = n4 * 2
  let n16 = n8 * 2

  assert (((n1 + n2 + n4) &&& (n2 + n4 + n8)) = n2 + n4)
  assert (((n1 + n2 + n4) ||| (n2 + n4 + n8)) = n1 + n2 + n4 + n8)
  assert (((n1 + n2 + n4) ^^^ (n2 + n4 + n8)) = n1 + n8)

  assert ((n1 <<< 0) = n1)
  assert ((n1 <<< 1) = n2)
  assert (((n1 + n2 + n8) <<< 1) = n2 + n4 + n16)

  assert ((n4 >>> 0) = n4)
  assert ((n4 >>> 1) = n2)
  assert ((n4 >>> 2) = n1)
  assert ((n4 >>> 3) = 0)
  assert (((n1 + n2 + n8) >>> 1) = 0 + n1 + n4)

  assert (~~~0 = 0xffffffff)
  assert (~~~0xffffffff = 0)
  assert (~~~0xca10cafe = 0x35ef3501)

let compareTest (n2: int) (n3: int) =
  assert (n2 = 2)
  assert (n3 <> 2)

  assert (n2 <= 2)
  assert (n2 <= n3)
  assert (n2 < n3)

  assert (n3 > n2)

  assert (compare n2 3 < 0)
  assert (compare n2 2 = 0)
  assert (compare n3 2 > 0)

let toIntTest () =
  assert (int 0 = 0)
  assert (int "42" = 42)
  assert (int "-1" + 1 = 0)

let main _ =
  literalTest ()
  hexLiteralTest ()
  arithmeticOperatorsTest 2 3 39
  bitOperatorsTest 1
  compareTest 2 3
  toIntTest ()
  0
